// @operational/components/lib/Internals/CardHeader.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const React = __importStar(require("react"));
const styled_1 = __importDefault(require("../utils/styled"));
const Container = styled_1.default("div")(({ theme }) => ({
    fontFamily: theme.font.family.main,
    fontSize: theme.font.size.body,
    fontWeight: theme.font.weight.medium,
    label: "cardheader",
    display: "flex",
    alignItems: "center",
    backgroundColor: theme.color.background.lighter,
    color: theme.color.text.lighter,
    flex: "0 0 auto",
    // This ensures that the card header text and card controls are placed in opposite corners.
    justifyContent: "space-between",
    height: theme.space.element * 2,
    padding: `0 ${theme.space.element}px`,
    lineHeight: 1,
    "& > :not(:first-child)": {
        fontSize: theme.font.size.fineprint,
        color: theme.color.text.lightest,
    },
    /**
     * Use case: External Links typically have <Icon/>s next to them.
     */
    "& a": {
        display: "inline-flex",
        alignItems: "center",
        textDecoration: "none",
    },
}));
const ActionsContainer = styled_1.default("div") `
  display: flex;
  align-items: center;
`;
const CardHeader = (_a) => {
    var { title, children, action } = _a, props = __rest(_a, ["title", "children", "action"]);
    return (React.createElement(Container, Object.assign({}, props),
        React.createElement("div", null, title || children),
        React.createElement(ActionsContainer, null, action)));
};
exports.default = CardHeader;
