// @operational/components/lib/Internals/Modal.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const React = __importStar(require("react"));
const ControlledModal_1 = __importDefault(require("./ControlledModal"));
class Modal extends React.Component {
    constructor() {
        super(...arguments);
        this.state = {
            options: {},
        };
        this.openModal = (options) => {
            this.setState({ options });
        };
        this.closeModal = () => {
            this.setState({ options: {} });
        };
    }
    render() {
        const isOpen = Boolean(this.state.options.body);
        return (React.createElement(React.Fragment, null,
            this.props.children(this.openModal),
            isOpen && (React.createElement(ControlledModal_1.default, { fullSize: this.state.options.fullSize, title: this.state.options.title, onClose: this.closeModal }, typeof this.state.options.body === "function"
                ? this.state.options.body(this.closeModal)
                : this.state.options.body))));
    }
}
exports.Modal = Modal;
exports.default = Modal;
