// @operational/components/lib/utils/mixins.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_emotion_1 = __importStar(require("react-emotion"));
const Tooltip_1 = require("../Tooltip/Tooltip");
const utils_1 = require("../utils");
exports.inputFocus = ({ theme, isError }) => ({
    outline: 0,
    border: "1px solid",
    borderColor: isError ? theme.color.error : theme.color.primary,
    boxShadow: `0 0 0 3px ${isError ? utils_1.lighten(theme.color.error, 60) : utils_1.lighten(theme.color.primary, 40)}`,
});
exports.Label = react_emotion_1.default("label")(({ fullWidth, theme, left }) => ({
    display: "inline-block",
    position: "relative",
    width: "100%",
    maxWidth: fullWidth ? "none" : "360px",
    marginRight: left ? theme.space.small : 0,
}));
exports.FormFieldControls = react_emotion_1.default("div")({
    position: "absolute",
    top: 3,
    right: 0,
});
exports.hoverTooltip = {
    [`.${Tooltip_1.dangerousTooltipContainerClassName}`]: {
        visibility: "hidden",
        pointerEvents: "none",
    },
    ":hover": {
        [`.${Tooltip_1.dangerousTooltipContainerClassName}`]: {
            visibility: "visible",
            pointerEvents: "all",
        },
    },
};
exports.FormFieldControl = react_emotion_1.default("div")(({ theme }) => (Object.assign({ cursor: "pointer", position: "relative", verticalAlign: "middle", display: "inline-block", width: "fit-content", marginLeft: theme.space.base, color: theme.color.text.lightest }, exports.hoverTooltip, { "& svg": {
        position: "relative",
    }, ":hover": {
        color: theme.color.text.default,
    } })));
exports.FormFieldError = react_emotion_1.default("div")(({ theme }) => ({
    fontSize: theme.font.size.fineprint,
    color: theme.color.error,
    padding: `${theme.space.base / 2}px ${theme.space.element}px`,
    marginBottom: 0,
    width: "100%",
    borderBottomLeftRadius: theme.borderRadius,
    borderBottomRightRadius: theme.borderRadius,
    border: `1px solid ${theme.color.error}`,
    borderTop: 0,
    transform: "translate3d(0, 100%, 0)",
    position: "absolute",
    backgroundColor: utils_1.lighten(theme.color.error, 60),
    boxShadow: `0px 3px 5px #d3d1d3`,
    // Nudge up just a little bit to look blended into the form
    bottom: 2,
    left: 0,
    zIndex: theme.zIndex.formFieldError,
}));
exports.floatIn = react_emotion_1.keyframes({
    from: {
        opacity: 0,
        transform: "translate3d(0, -6px, 0)",
    },
    to: {
        opacity: 1,
        transform: "translate3d(0, 0, 0)",
    },
});
exports.resetTransform = react_emotion_1.keyframes({
    to: {
        transform: "none",
    },
});
exports.spin = react_emotion_1.keyframes({
    from: {
        transform: "rotate(0deg)",
    },
    to: {
        transform: "rotate(359deg)",
    },
});
