// @operational/components/lib/PageAreas/PageAreas.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const React = __importStar(require("react"));
const styled_1 = __importDefault(require("../utils/styled"));
const Container = styled_1.default("div")(props => {
    const gridTemplateColumns = {
        main: "auto",
        "main side": "auto 280px",
        "side main": "280px auto",
    }[props.areas || "main"];
    return {
        gridTemplateColumns,
        display: "grid",
        alignItems: "flex-start",
        gridTemplateAreas: `"${props.areas || "main"}"`,
        gridGap: props.theme.space.element,
        maxWidth: props.fill_ ? "none" : 1150,
        minWidth: 800,
        width: "100%",
        "> div": {
            height: "100%",
        },
        ":not(:last-child)": {
            marginBottom: props.theme.space.element,
        },
    };
});
const PageAreas = (_a) => {
    var { fill } = _a, props = __rest(_a, ["fill"]);
    return React.createElement(Container, Object.assign({}, props, { fill_: fill }));
};
exports.default = PageAreas;
