// @operational/components/lib/Table/Table.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const React = __importStar(require("react"));
const ActionMenu_1 = __importDefault(require("../ActionMenu/ActionMenu"));
const Icon_1 = __importDefault(require("../Icon/Icon"));
const Small_1 = __importDefault(require("../Typography/Small"));
const styled_1 = __importDefault(require("../utils/styled"));
const Container = styled_1.default("table")(({ theme }) => ({
    width: "100%",
    backgroundColor: theme.color.white,
    textAlign: "left",
    borderCollapse: "collapse",
    fontSize: theme.font.size.small,
    fontFamily: theme.font.family.main,
}));
const Tr = styled_1.default("tr")(({ hover, theme }) => (Object.assign({ height: 50 }, (hover
    ? {
        ":hover": {
            backgroundColor: theme.color.background.lighter,
            cursor: "pointer",
        },
    }
    : {}))));
const Th = styled_1.default("th")(({ theme }) => ({
    verticalAlign: "bottom",
    borderBottom: `1px solid ${theme.color.separators.default}`,
    color: theme.color.text.lightest,
    paddingBottom: theme.space.base,
    "&:first-child": {
        paddingLeft: theme.space.small,
    },
}));
const Td = styled_1.default("td")(({ theme }) => ({
    verticalAlign: "middle",
    borderBottom: `1px solid ${theme.color.separators.default}`,
    color: theme.color.text.default,
    "&:first-child": {
        paddingLeft: theme.space.small,
    },
}));
const Actions = styled_1.default(Td)(({ theme }) => ({
    textAlign: "right",
    paddingRight: theme.space.small,
    /**
     * We use opacity here instead of display: none; or
     * visibility: hidden; because both mess with
     * the box model of the Td while opacity does not.
     */
    opacity: 0,
    "tr:hover &, :hover": {
        opacity: 1,
    },
    "& > div": {
        display: "inline-flex",
    },
}));
const IconCell = styled_1.default(Td) `
  width: 40px;
  padding: ${props => props.theme.space.base}px;
  color: ${props => props.theme.color.text.lightest};
`;
const ActionLabel = styled_1.default(Small_1.default) `
  color: ${props => props.theme.color.primary};
  margin: 0;
  display: block;
`;
const EmptyView = styled_1.default(Td)(({ theme }) => ({
    color: theme.color.text.default,
    height: 50,
    lineHeight: "50px",
    textAlign: "center",
}));
function Table(_a) {
    var { data = [], columns, onRowClick, rowActionName, rowActions, icon, iconColor } = _a, props = __rest(_a, ["data", "columns", "onRowClick", "rowActionName", "rowActions", "icon", "iconColor"]);
    const standardizedColumns = Boolean(columns[0]) && typeof columns[0] === "string"
        ? columns.map(columnName => ({
            heading: columnName,
            cell: (dataEntry) => dataEntry[columnName],
        }))
        : columns;
    const hasIcons = Boolean(data[0]) && Boolean(icon) && Boolean(icon(data[0]));
    return (React.createElement(Container, Object.assign({}, props),
        React.createElement("thead", null,
            React.createElement(Tr, null,
                hasIcons && React.createElement(Th, { key: "-1" }),
                standardizedColumns.map((column, columnIndex) => React.createElement(Th, { key: columnIndex }, column.heading)),
                Boolean(rowActions || (onRowClick && rowActionName)) && React.createElement(Th, { key: "infinity" }))),
        React.createElement("tbody", null, data.length ? (data.map((dataEntry, dataEntryIndex) => {
            const rowAction = (() => {
                if (!rowActions) {
                    return null;
                }
                const dataEntryRowActions = rowActions(dataEntry);
                return (React.createElement(Actions, { onClick: (ev) => {
                        // Table row click should not trigger if this action menu is manipulated
                        ev.stopPropagation();
                    } }, Array.isArray(dataEntryRowActions) ? (React.createElement(ActionMenu_1.default, { items: dataEntryRowActions })) : (dataEntryRowActions)));
            })();
            return (React.createElement(Tr, { hover: Boolean(data), key: dataEntryIndex, onClick: () => {
                    if (onRowClick) {
                        onRowClick(dataEntry, dataEntryIndex);
                    }
                } },
                hasIcons && (React.createElement(IconCell, null,
                    React.createElement(Icon_1.default, { name: icon(dataEntry), color: iconColor && iconColor(dataEntry) }))),
                standardizedColumns.map((column, columnIndex) => (React.createElement(Td, { key: columnIndex }, column.cell(dataEntry, dataEntryIndex)))),
                rowAction,
                onRowClick &&
                    rowActionName && (React.createElement(Actions, null,
                    React.createElement(ActionLabel, null, rowActionName)))));
        })) : (React.createElement(Tr, null,
            React.createElement(EmptyView, { colSpan: columns.length }, "There are no records available"))))));
}
exports.default = Table;
