// @operational/components/lib/CardColumn/CardColumn.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const React = __importStar(require("react"));
const styled_1 = __importDefault(require("../utils/styled"));
const Container = styled_1.default("div")(({ theme, contentRight, flexColumn }) => (Object.assign({ label: "card-column", height: "min-content", minWidth: 280 / 2, padding: theme.space.element / 2, flex: "1 0", " img": {
        maxWidth: "100%",
    }, textAlign: contentRight ? "right" : "left" }, (flexColumn
    ? {
        display: "flex",
        flexDirection: "column",
        justifyContent: contentRight ? "flex-end" : "flex-start",
    }
    : {}))));
const Title = styled_1.default("div")(({ theme }) => ({
    fontFamily: theme.font.family.main,
    fontWeight: theme.font.weight.bold,
    color: theme.color.text.dark,
    fontSize: theme.font.size.body,
    borderBottom: `1px solid ${theme.color.separators.default}`,
    paddingBottom: theme.space.small,
    marginBottom: theme.space.content,
}));
const CardColumn = (_a) => {
    var { title, children } = _a, props = __rest(_a, ["title", "children"]);
    return (React.createElement(Container, Object.assign({}, props),
        title && React.createElement(Title, null, title),
        children));
};
exports.default = CardColumn;
