// @operational/components/lib/ButtonGroup/ButtonGroup.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const React = __importStar(require("react"));
const styled_1 = __importDefault(require("../utils/styled"));
const Container = styled_1.default("div")({
    label: "buttongroup",
    "& > button": {
        margin: 0,
    },
    "& > button:not(:first-child)": {
        // To avoid overlapping borders
        marginLeft: -1,
        borderLeft: 0,
        borderTopLeftRadius: 0,
        borderBottomLeftRadius: 0,
    },
    "& > button:not(:last-child)": {
        borderTopRightRadius: 0,
        borderBottomRightRadius: 0,
    },
});
const ButtonGroup = (_a) => {
    var { children } = _a, props = __rest(_a, ["children"]);
    return (React.createElement(Container, Object.assign({}, props), children));
};
exports.default = ButtonGroup;
