// @operational/components/lib/ActionMenu/ActionMenu.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const React = __importStar(require("react"));
const ContextMenu_1 = __importDefault(require("../ContextMenu/ContextMenu"));
const Icon_1 = __importDefault(require("../Icon/Icon"));
const utils_1 = require("../utils");
const styled_1 = __importDefault(require("../utils/styled"));
const width = 144;
const Container = styled_1.default("div")(({ theme }) => ({
    width,
    height: 35,
    /**
     * `textAlign` is set explicitly for when a parent sets a text-align to right-position this container,
     * leaving its content left-aligned.
     */
    textAlign: "left",
    padding: `0 ${theme.space.content}px`,
    backgroundColor: theme.color.white,
    color: theme.color.primary,
    fontWeight: theme.font.weight.medium,
    borderRadius: theme.borderRadius,
    display: "flex",
    alignItems: "center",
    justifyContent: "space-between",
    userSelect: "none",
    boxShadow: `0 0 0 1px ${theme.color.separators.light}`,
    "&:hover": {
        backgroundColor: utils_1.darken(theme.color.white, 5),
    },
}));
const TitleContainer = styled_1.default("p")({
    width: 90,
    whiteSpace: "nowrap",
    overflow: "hidden",
    textOverflow: "ellipsis",
});
const ActionMenu = props => (React.createElement(ContextMenu_1.default, Object.assign({}, props, { items: props.items, width: width, condensed: true, embedChildrenInMenu: true }), isOpen => (React.createElement(Container, null,
    React.createElement(TitleContainer, null, props.title),
    React.createElement(Icon_1.default, { name: isOpen ? "ChevronUp" : "Menu" })))));
ActionMenu.defaultProps = {
    title: "Actions",
};
exports.default = ActionMenu;
